modules/engine/src/main/scala/com/gsk/kg/engine/Multiset.scala

```diff
@@ -370,6 +362,88 @@ object Multiset {
       .createDataFrame(result.toJavaRDD, resultSchema)
   }
 
+  /** This method performs a inner join between graphs by merging the graph columns of each dataframes into an array
+    * column, if they have same graph then the graph is preserved, if not it is added as if in the default graph. Eg:
+    * Initial dataframes:
+    * l.dataframe = List(
+    *   ("a", "graph1"),
+    *   ("b", "graph1"),
+    *   ("c", "graph1")
+    * ).toDF("?x", "*g")
+    * r.dataframe = List(
+    *   ("b", "graph1"),
+    *   ("c", "graph2"),
+    *   ("d", "graph2")
+    * ).toDF("?x", "*g")
+    *
+    * Step1:
+    * step1Dataframe = List(
+    *   ("b", "graph1", "graph1"),
+    *   ("c", "graph1", "graph2")
+    * ).toDF("?x", "*l", "*r")
+    *
+    * Step2:
+    * step2Dataframe = List(
+    *   ("b", List("graph1", "graph1")),
+    *   ("c", List("graph1", "graph2"))
+    * ).toDF("?x", "*g")
+    *
+    * Final:
+    * final = List(
+    *  ("b", "graph1"),
+    *  ("c", ""),
+    * ).toDF("?x", "*g")
+    *
+    * IMPORTANT!: This method could have some performance issues as it traverses entire dataframe, and should be revisited
+    * for performance improvements.
+    * @param l
+    * @param r
+    * @return
+    */
+  def innerJoinWithGraphsColumn(l: DataFrame, r: DataFrame): DataFrame = {
+    val leftGraphCol  = "*l"
+    val rightGraphCol = "*r"
+
+    val renamedLeft =
+      l.as("l").withColumnRenamed(GRAPH_VARIABLE.s, leftGraphCol)
+    val renamedRight =
+      r.as("r").withColumnRenamed(GRAPH_VARIABLE.s, rightGraphCol)
+
+    val intersect = renamedLeft.columns intersect renamedRight.columns
+    val innerJoin = renamedLeft.join(renamedRight, intersect, "inner")
+
+    // Merges the *l and *r columns into a *g column with Array(l, r)
+    val innerWithMergedGraphColunns = innerJoin
+      .withColumn(GRAPH_VARIABLE.s, array(leftGraphCol, rightGraphCol))
+      .drop(leftGraphCol, rightGraphCol)
+
+    // Generates a schema for the final DF (needed for the flatMap)
+    val resultSchema = innerWithMergedGraphColunns
+      .drop(leftGraphCol, rightGraphCol)
```

Comment:
Do we need to drop the columns twice?
```suggestion

    // Generates a schema for the final DF (needed for the flatMap)
    val resultSchema = innerWithMergedGraphColunns
      .drop(leftGraphCol, rightGraphCol)
```

Reply:
In this case we are generating an schema to use later in the .flatMap as RowEncoder, so we construct the schema based on the input Dataset, but the output Dataset from flatmap doesn't have that columns. Do you think there is a better way of generating the output schema?

Reply:
Oh no, I just meant that we had already dropped leftgraphcol and rightgraphcol from innerWithMergedGraphColunns

Reply:
oohhh I see, good catch mate!